### PR TITLE
P1 - FE - Asset Detail page (ADP) [24] [scoping reserved]

### PR DESCRIPTION
**Pull request by Space-zee**

### CODE REVIEW

**Comment by Space-zee** [comment]
https://app.clickup.com/t/33yxp9c

File: frontend/app/features/assets/controllers/AssetsDetailsController.module.scss [hunk 1]
```diff
@@ -1,5 +1,64 @@
...
+.assetInvestDetails{
+  padding: 24px;
+}
+.loader{
+  min-height: 200px;
+  z-index: 10;
+}
+
+
+@media screen and (max-width: 480px) {
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/controllers/AssetsDetailsController.module.scss:46]: why screen only?

File: frontend/app/features/assets/models/tokenBuyValidator.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import defaultMessagesEn from "elegant-validator/src/defaultMessagesEn";
+
+
+const ev = new ElegantValidator(defaultValidators, defaultMessagesEn)
+
+const schema= {
+  amount: ['required'],
+}
+export const tokenAmountValidator = {
+  isAmountValid(formData: {amount: number}, tokensLeft: number, userBalance: number, tokenPrice: number) {
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/models/tokenBuyValidator.ts:12]: there is much better way here to validate, but let's keep as it is for now

File: frontend/app/features/assets/views/AssetInfo/Tabs/Tabs/index.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        <Tab id={'financials'} title={'Financials'} activeTab={activeTab} setActiveTab={setActiveTab}/>
+        <Tab id={'documents'} title={'Documents'} activeTab={activeTab} setActiveTab={setActiveTab}/>
+        <Tab id={'buyingProcess'} title={'Buying Process'} activeTab={activeTab} setActiveTab={setActiveTab}/>
+        <Tab id={'market'} title={'Market'} activeTab={activeTab} setActiveTab={setActiveTab}/>
+      </div>
+      <div className={s.outlet}>
+        <TabContent id="details" activeTab={activeTab}>
+          <div className={s.detailsIcons}>
+            <IconWrapper classname={s.detailsIconElem}>
+              <Image src={bedAIcon} alt={'bedAIcon'}/>
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/AssetInfo/Tabs/Tabs/index.tsx:35]: you can simply put src="/public/feature/assets/bed.svg" 
isn't it good enough?

File: frontend/app/features/assets/views/InvestInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+type Props = {
+  balance: number;
+  tokensLeft: number;
+  tokenPrice: number;
+  setInvestAmount: (value: number) => void;
+  setValidationInvestError: (value: string) => void
+  validationInvestError: string;
+}
+export const InvestInput = (props: Props) => {
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/InvestInput/index.tsx:14]: 1. that's what I ment to NOT put validation into form nor inputs
2. this is not an input - this is InvestForm component. Just validation must be in model (or at least in contoller)

File: frontend/app/features/assets/views/Location/index.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      <p className={s.header}>
+        Location
+      </p>
+       <GoogleMapReact
+         bootstrapURLKeys={{key: ''}}
+         defaultCenter={location}
+         defaultZoom={20}
+       >
+         <LocationPin
+           // @ts-ignore
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/Location/index.tsx:23]: please do not abuse ts-ignore

File: frontend/app/styles/globalVariables.css [hunk 6]
```diff
@@ -30,7 +30,7 @@
   --border-radius_large: 10px;
 
   --max-width_desktop: 1200px;
-  --max-width_mobile: 320px;
+  --max-width_mobile: 480px;
```
Comment (ruslanpalagin) [inline frontend/app/styles/globalVariables.css:33]: have you checked all the pages to make sure you did not break anything?

File: frontend/app/hooks/useWindowSize.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  sm: boolean;
+  md: boolean;
+  lg: boolean;
+}
+
+export const useWindowSize = (): ScreenSizeInterface => {
+  const [windowSize, setWindowSize] = useState(0);
+  useEffect(() => {
+    function handleResize() {
+      setWindowSize(window.innerWidth);
```
Comment (ruslanpalagin) [inline frontend/app/hooks/useWindowSize.ts:21]: it's bad practice to use setState on each event. 
you should add throttle/debounce in such a case...

up to you - we can optimize this later

**Review by ruslanpalagin — CHANGES_REQUESTED** [review]

File: frontend/app/features/assets/views/Location/index.tsx (hunk 5 above)
Comment (Space-zee) [reply frontend/app/features/assets/views/Location/index.tsx:23]: +

File: frontend/app/features/assets/views/InvestInput/index.tsx (hunk 4 above)
Comment (Space-zee) [reply frontend/app/features/assets/views/InvestInput/index.tsx:14]: +

File: frontend/app/features/assets/controllers/AssetsDetailsController.module.scss (hunk 1 above)
Comment (Space-zee) [reply frontend/app/features/assets/controllers/AssetsDetailsController.module.scss:46]: +

File: frontend/app/features/assets/models/tokenBuyValidator.ts (hunk 2 above)
Comment (Space-zee) [reply frontend/app/features/assets/models/tokenBuyValidator.ts:12]: +

File: frontend/app/hooks/useWindowSize.ts (hunk 7 above)
Comment (Space-zee) [reply frontend/app/hooks/useWindowSize.ts:21]: +

File: frontend/app/features/assets/views/AssetInvest/index.tsx [hunk 8]
```diff
@@ -35,16 +34,23 @@ export const AssetInvest = ({
                               validationInvestError,
                               validateInvestInput
                             }: InvestBlockProps) => {
+  const router = useRouter();
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/AssetInvest/index.tsx:37]: Please avoid usage router in views.
Where to navigate - is a business logic and should be managed by controller or model

File: frontend/app/features/assets/views/InvestInput/index.tsx [hunk 9]
```diff
@@ -1,28 +1,34 @@
...
+  const {tokensLeft, setInvestAmount, validationInvestError, validateInput, investAmount} = props
   const setMax = () => {
-    // @ts-ignore
-    inputRef.current.value = balance
+    setInvestAmount(tokensLeft.toString());
   }
+  const onlyNumbers = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const regExp = /^[0-9\b]+$/;
+    if (e.target.value === '' || regExp.test(e.target.value)) {
+      setInvestAmount(e.target.value);
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/InvestInput/index.tsx:20]: Please use views naming: 
 setInvestAmount -> onAmountChange

File: frontend/package.json [hunk 10]
```diff
@@ -17,6 +17,7 @@
     "@types/babel__core": "^7.1.20",
     "@types/google-map-react": "^2.1.7",
     "@types/lodash.debounce": "^4.0.7",
+    "@types/next": "^9.0.0",
```
Comment (ruslanpalagin) [inline frontend/package.json:20]: Cool, thanks

**Review by ruslanpalagin — CHANGES_REQUESTED** [review]

**Review by ruslanpalagin — APPROVED** [review]
Great progress, thanks!

**Pull request merged**